src/recordlinker/linkage/models.py

```diff
@@ -1,58 +1,182 @@
+    the algorithm.  By defining them all upfront, we give the user flexibility in
+    adjusting their algorithm configuration without having to reload the data.
+
+    **HERE BE DRAGONS**: IN A PRODUCTION SYSTEM, THESE ENUMS SHOULD NOT BE CHANGED!!!
+    """
 
-class BlockingKey(Base):
-    __tablename__ = "mpi_blocking_key"
+    BIRTHDATE = 1, "Date of Birth"
+    MRN = 2, "Last 4 chars of MRN"
+    SEX = 3, "Sex"
+    ZIP = 4, "Zip Code"
+    FIRST_NAME = 5, "First 4 chars of First Name"
+    LAST_NAME = 6, "First 4 chars of Last Name"
+
+    def __init__(self, id: int, description: str):
+        self.id = id
+        self.description = description
+
+    def to_value(self, data: dict) -> set[str]:
+        """
+        Given a data dictionary of Patient PII data, return a set of all
+        possible values for this Key.  Many Keys will only have 1 possible value,
+        but some (like first name) could have multiple values.
+        """
+        vals: set[str] = set()
+        if self == BlockingKey.BIRTHDATE:
+            vals.update(self._extract_birthdate(data))
+        if self == BlockingKey.MRN:
+            vals.update(self._extract_mrn_last4(data))
+        if self == BlockingKey.SEX:
+            vals.update(self._extract_sex(data))
+        if self == BlockingKey.ZIP:
+            vals.update(self._extract_zipcode(data))
+        if self == BlockingKey.FIRST_NAME:
+            vals.update(self._extract_first_name_first4(data))
+        if self == BlockingKey.LAST_NAME:
+            vals.update(self._extract_last_name_first4(data))
+
+        # remove all empty strings from the set, we never want to block on these
+        vals.discard("")
+        return vals
+
+    # FIXME: some of the type/isinstance checks below can be removed after we create
+    # a serializer to capture the input data, as type dict is too flexible
+    def _extract_birthdate(self, data: dict) -> typing.Iterator[str]:
+        val = data.get("birthdate")
+        if val:
+            if not isinstance(val, (datetime.date, datetime.datetime)):
+                # if not an instance of date or datetime, try to parse it
+                val = dateutil.parser.parse(str(val))
+            yield val.strftime("%Y-%m-%d")
+
+    def _extract_mrn_last4(self, data: dict) -> typing.Iterator[str]:
+        if data.get("mrn"):
+            yield data["mrn"].strip()[-4:]
+
+    def _extract_sex(self, data: dict) -> typing.Iterator[str]:
+        if data.get("sex"):
+            val = str(data["sex"]).lower().strip()
+            if val in ["m", "male"]:
+                yield "m"
+            elif val in ["f", "female"]:
+                yield "f"
+            else:
+                yield "u"
+
+    def _extract_zipcode(self, data: dict) -> typing.Iterator[str]:
+        val = data.get("address")
+        if isinstance(val, (list, set)):
+            for address in val:
+                if isinstance(address, dict):
```

Comment:
Should these `extract...` functions be returning something if the `val` is NOT a `list` or `set`? Is it always expected that we would have to go multiple layers deep into the `data` parameter to get the value we want to return? I'm thinking like what would return from this function if any of the `isinstance` checks failed.